src/metabase/models/session.clj

```diff
@@ -29,3 +39,68 @@
   [{anti-csrf-token :anti_csrf_token, :as session}]
   (let [session-type (if anti-csrf-token :full-app-embed :normal)]
     (assoc session :type session-type)))
+
+(defn maybe-send-login-from-new-device-email
+  "If set to send emails on first login from new devices, that is the case, and its not the users first login, send an
+  email from a separate thread."
+  [history-record]
+  (when (and (login-history/send-email-on-first-login-from-new-device)
+             (login-history/first-login-on-this-device? history-record)
+             (not (login-history/first-login-ever? history-record)))
+    ;; if there's an existing open connection (and there seems to be one, but I'm not 100% sure why) we can't try to use
+    ;; it across threads since it can close at any moment! So unbind it so the future can get its own thread.
+    (binding [t2.conn/*current-connectable* nil]
+      (future
+        ;; off thread for both IP lookup and email sending. Either one could block and slow down user login (#16169)
+        (try
+          (let [[info] (login-history/human-friendly-infos [history-record])]
+            (messages/send-login-from-new-device-email! info))
+          (catch Throwable e
+            (log/error e "Error sending 'login from new device' notification email")))))))
+
+(def ^:private CreateSessionUserInfo
+  [:map
+   [:id ms/PositiveInt]
+   [:last_login :any]])
+
+(def SessionSchema
+  "Schema for a Session."
+  [:and
+   [:map-of :keyword :any]
+   [:map
+    [:id uuid?]
+    [:type [:enum :normal :full-app-embed]]]])
+
+(defmulti create-session!
+  "Generate a new Session for a User. `session-type` is currently either `:password` (for email + password login) or
+  `:sso` (for other login types). Returns the newly generated Session."
+  {:arglists '(^java.util.UUID [session-type user device-info])}
```

Comment:
I see that this was copied from the previous implementation, but I don't really get the type hint here (the `^java.util.UUID` bit). It looks like we're type hinting the whole vector, but... it's not a UUID. I wonder if this was a mistake at some point?